funcs/func_hangupcause.c

```diff
@@ -139,25 +141,50 @@ static int hangupcause_read(struct ast_channel *chan, const char *cmd, char *dat
 		return -1;
 	}
 
-	ast_channel_lock(chan);
-	cause_code = ast_channel_dialed_causes_find(chan, args.channel);
-	ast_channel_unlock(chan);
+//	ast_log(LOG_ERROR, "Cause for channel '%s'\n", args.channel);
 
-	if (!cause_code) {
-		ast_log(LOG_WARNING, "Unable to find information for channel %s\n", args.channel);
+	if (!(causelist = ast_str_create(128))) {
+		ast_log(LOG_ERROR, "Unable to allocate dynamic string buffer\n");
 		return -1;
 	}
 
+	ast_channel_lock(chan);
+	if (!strcmp(args.type, "tech_extended")) {
+		struct ao2_iterator *cause_codes;
+		cause_codes = ast_channel_dialed_causes_find_multiple(chan, args.channel);
+		while ((cause_code = ao2_iterator_next(cause_codes))) {
+			if (!cause_code->cause_extended) {
+				ao2_ref(cause_code, -1);
+				continue;
+			}
+			ast_str_append(&causelist, 0, "%s%s", (ast_str_strlen(causelist) ? "," : ""), cause_code->code);
+			ao2_ref(cause_code, -1);
+		}
+		ao2_iterator_destroy(cause_codes);
+	} else {
+		cause_code = ast_channel_dialed_causes_find(chan, args.channel);
+		if (!cause_code) {
+			ast_log(LOG_WARNING, "Unable to find information for channel '%s'\n", args.channel);
+			ast_channel_unlock(chan);
+			return -1;
+		}
+	}
+	ast_channel_unlock(chan);
+
 	if (!strcmp(args.type, "ast")) {
 		ast_copy_string(buf, ast_cause2str(cause_code->ast_cause), len);
 	} else if (!strcmp(args.type, "tech")) {
 		ast_copy_string(buf, cause_code->code, len);
+	} else if (!strcmp(args.type, "tech_extended")) {
+		ast_copy_string(buf, ast_str_buffer(causelist), len);
 	} else {
 		ast_log(LOG_WARNING, "Information type not recognized (%s)\n", args.type);
 		res = -1;
 	}
 
-	ao2_ref(cause_code, -1);
+	if (cause_code) {
```

Comment:
There's an ao2_cleanup for this, you just need to replace the ao2_ref call with it